base/channels.jl

```diff
@@ -212,11 +212,52 @@ function close(c::Channel, @nospecialize(excp::Exception))
     nothing
 end
 
-# Use acquire here to pair with release store in `close`, so that subsequent `isready` calls
-# are forced to see `isready == true` if they see `isopen == false`. This means users must
-# call `isopen` before `isready` if you are using the race-y APIs (or call `iterate`, which
-# does this right for you).
-isopen(c::Channel) = ((@atomic :acquire c.state) === :open)
+"""
+    isopen(c::Channel)
+Determines whether a [`Channel`](@ref) is open for new [`put!`](@ref) operations.
+Notice that a `Channel`` can be closed and still have
+buffered elements which can be consumed with [`take!`](@ref).
+
+# Examples
+
+Buffered channel with task:
+```jldoctest
+julia> c = Channel(ch -> put!(ch, 1), 1);
+
+julia> isopen(c)
+false # channel is closed to new `put!`s
+
+# even though the channel is closed, it still contains elements
+julia> isready(c)
+true
+
+juiia> take!(c)
+1
+
+julia> isready(c)
+false
+```
+
+Unbuffered channel:
+```jldoctest
+julia> c = Channel{Int}()
+
+julia> isopen(c)
+true
+
+julia> close(c)
+
+julia> isopen(c)
+false
```

Comment:
```suggestion
```

Reply:
Really? I thought, once an unbuffered channel is closed, all concurrent `put!`s error?

```julia
julia> ch = Channel()
Channel{Any}(0) (empty)

julia> t = @async put!(ch, 1)
Task (runnable) @0x000002028d9f1f10

julia> isready(ch)
true

julia> close(ch)

julia> isopen(ch)
false

julia> isready(ch)
false

julia> fetch(t)
ERROR: TaskFailedException
Stacktrace:
 [1] wait
   @ .\task.jl:359 [inlined]
 [2] fetch(t::Task)
   @ Base .\task.jl:379
 [3] top-level scope
   @ REPL[20]:1

    nested task error: InvalidStateException: Channel is closed.
    Stacktrace:
     [1] try_yieldto(undo::typeof(Base.ensure_rescheduled))
       @ Base .\task.jl:945
     [2] wait()
       @ Base .\task.jl:1009
     [3] wait(c::Base.GenericCondition{ReentrantLock}; first::Bool)
       @ Base .\condition.jl:130
     [4] wait
       @ .\condition.jl:125 [inlined]
     [5] put_unbuffered(c::Channel{Any}, v::Int64)
       @ Base .\channels.jl:387
     [6] put!(c::Channel{Any}, v::Int64)
       @ Base .\channels.jl:342
     [7] (::var"#17#18")()
       @ Main .\REPL[15]:1
```

Reply:
Ah, okay, I didn't know that. I guess that makes sense though

Reply:
This is why I keeping hitting this issue. Since the behavior of unbuffered and buffered channels is subtly different, I'll add a buffer to a channel and introduce silent bugs elsewhere! Thank you for the review!